tests/test_class.py

```diff
@@ -323,3 +323,8 @@ def test_non_final_final():
         class PyNonFinalFinalChild(m.IsNonFinalFinal):
             pass
     assert str(exc_info.value).endswith("is not an acceptable base type")
+
+
+def test_base_and_derived_nested_scope():
+    assert issubclass(m.DerivedWithNested, m.BaseWithNested)
+    assert m.DerivedWithNested.Nested != m.BaseWithNested.Nested
```

Comment:
I consider this a very weak assert, because there could be bugs or even features making this unequal by chance.
Is there a simple way to make this more conclusive? E.g. modeling #1624 more closely? Or adding a method `int foo() { return 1; }` for base nested and `return 2` for the derived nested?

Reply:
Methods should not be affected by this, I believe? I've never seen or had any complaints where you can't override methods in a derived class.

The main test is actually that the module imports, because in #1624, pybind11 refuses to override the `Nested` class in the base class. So if the import succeeds and the two `Nested` classes are different, we don't regress on this PR, I believe. (The test fails without the fixes.)

> I consider this a very weak assert, because there could be bugs or even features making this unequal by chance.

Should I `assert m.DerivedWithNested.Nested is not m.BaseWithNested.Nested`? I guess that would result in the same check on the two nested classes being distinct, but that can't be overridden by `__eq__` in the metaclass.

Reply:
I guess I could still add some static property to both classes, though.

Reply:
Up to you, if you feel this is sufficiently strong I'm fine, but in my experience trying to be stingy in test code will bite you sooner or later. The basic idea: find a reasonably simple way to be absolutely sure some other bug or misunderstanding cannot fool us here. Often that means you're not purely testing what you actually want to test, but I see that as a safety cushion.

Reply:
I see what you mean.
In this case, I'm still not convinced it will catch something more, but it can't really hurt either, so ... done :-)

I still want to get a look at @bstaletic's links as well, to see if accessing `__dict__` is really the correct solution.

Reply:
That looks like a great safety cushion to me: just 4 more lines, difficult to image that we could still get fooled.

Reply:
Right, thanks for the review! :-)